ejector/data_api_signing_rate_lookup.go

```diff
@@ -0,0 +1,316 @@
+	timeSpan time.Duration,
+	quorums []core.QuorumID,
+	version ProtocolVersion,
+	omitPerfectSigners bool,
+) ([]*validator.ValidatorSigningRate, error) {
+	switch version {
+	case ProtocolVersionV1:
+		if !omitPerfectSigners {
+			srl.logger.Warn(
+				"omitPerfectSigners flag is ignored for ProtocolVersionV1, will never return perfect signers")
+		}
+		return srl.getV1SigningRates(timeSpan, quorums)
+	case ProtocolVersionV2:
+		return srl.getV2SigningRates(timeSpan, quorums, omitPerfectSigners)
+	default:
+		return nil, fmt.Errorf("unsupported protocol version: %d", version)
+	}
+}
+
+// Look up signing rates for v1.
+func (srl *dataApiSigningRateLookup) getV1SigningRates(
+	timeSpan time.Duration,
+	quorums []core.QuorumID,
+) ([]*validator.ValidatorSigningRate, error) {
+
+	quorumSet := make(map[core.QuorumID]struct{})
+	for _, q := range quorums {
+		quorumSet[q] = struct{}{}
+	}
+
+	now := time.Now()
+
+	path := "api/v1/metrics/operator-nonsigning-percentage"
+	urlStr, err := url.JoinPath(srl.url, path)
+	if err != nil {
+		return nil, fmt.Errorf("error joining URL path with %s and %s: %w", srl.url, path, err)
+	}
+	url, err := url.Parse(urlStr)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing URL: %w", err)
+	}
+	// add query parameters
+	q := url.Query()
+	// end: datetime formatted in "2006-01-02T15:04:05Z"
+	q.Set("end", now.UTC().Format("2006-01-02T15:04:05Z"))
+	// interval: lookback window in seconds
+	q.Set("interval", strconv.Itoa(int(timeSpan.Seconds())))
+	url.RawQuery = q.Encode()
+	// Very verbose, enable for debugging if needed.
+	// srl.logger.Debug("making request to DataAPI", "url", url.String())
+
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating HTTP request: %w", err)
+	}
+
+	resp, err := srl.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error sending HTTP request: %w", err)
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response body: %w", err)
+	}
+	// Very verbose, enable for debugging if needed.
+	// srl.logger.Info("Received response", "responseBody", string(respBody))
+
+	if resp.StatusCode != http.StatusOK {
+		var errResp dataapi.ErrorResponse
+		err = json.Unmarshal(respBody, &errResp)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing error response: %w", err)
+		}
+		return nil, fmt.Errorf(
+			"error response (%d) from dataapi: %s",
+			resp.StatusCode,
+			errResp.Error,
+		)
+	}
+
+	var response dataapi.OperatorsNonsigningPercentage
+	err = json.NewDecoder(strings.NewReader(string(respBody))).Decode(&response)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing response body: %w", err)
+	}
+
+	// Use a map to combine results from multiple quorums.
+	signingRateMap := make(map[core.OperatorID]*validator.ValidatorSigningRate)
+
+	for _, data := range response.Data {
+
+		if len(quorumSet) > 0 {
```

Comment:
Should you be asserting that quorum set isn't empty at the very beginning of the method? Seems the whole thing is pointless if no quorums are requested

(any changes here should be mirrored for the v2 method as well)

Reply:
I see now looking at the interface file that empty quorum set has a special meaning. Consider noting that somewhere in this file, since it seems like a bug at first glance